// !DIAGNOSTICS: -UNUSED_EXPRESSION
// !WITH_NEW_INFERENCE

import Obj.ext
import A.Companion.ext2

object Obj {
    val String.ext: String get() = this
}

class A {
    companion object {
        val String.ext2: String get() = this
    }
}

fun test() {
    String::<!EXTENSION_IN_CLASS_REFERENCE_NOT_ALLOWED!>ext<!>
    <!NI;TYPE_MISMATCH!>Obj::<!NI;EXTENSION_IN_CLASS_REFERENCE_NOT_ALLOWED, OI;UNRESOLVED_REFERENCE_WRONG_RECEIVER!>ext<!><!>

    String::<!EXTENSION_IN_CLASS_REFERENCE_NOT_ALLOWED!>ext2<!>
    <!NI;TYPE_MISMATCH!>A.Companion::<!NI;EXTENSION_IN_CLASS_REFERENCE_NOT_ALLOWED, OI;UNRESOLVED_REFERENCE_WRONG_RECEIVER!>ext2<!><!>
    <!NI;TYPE_MISMATCH!>A::<!NI;EXTENSION_IN_CLASS_REFERENCE_NOT_ALLOWED, OI;UNRESOLVED_REFERENCE_WRONG_RECEIVER!>ext2<!><!>

    A::<!UNRESOLVED_REFERENCE!>foo<!>
    A::<!UNRESOLVED_REFERENCE!>bar<!>
}
