// !WITH_NEW_INFERENCE
// StackOverflow
val p = <!TYPECHECKER_HAS_RUN_INTO_RECURSIVE_PROBLEM!>::<!NI;DEBUG_INFO_MISSING_UNRESOLVED!>p<!><!>

fun foo() = <!TYPECHECKER_HAS_RUN_INTO_RECURSIVE_PROBLEM!>::<!NI;DEBUG_INFO_MISSING_UNRESOLVED!>foo<!><!>